internals/api/credential.go

```diff
@@ -29,39 +39,132 @@ type CredentialType string
+	Fingerprint *string         `json:"fingerprint"`
+	Name        *string         `json:"name,omitempty"`
+	Verifier    []byte          `json:"verifier"`
+	Proof       interface{}     `json:"proof"`
 }
 
-// Validate validates the request fields.
-func (req CreateCredentialRequest) Validate() error {
-	if req.Fingerprint == "" {
-		return ErrInvalidFingerprint
+// UnmarshalJSON converts a JSON representation into a CreateCredentialRequest with the correct Proof.
+func (req *CreateCredentialRequest) UnmarshalJSON(b []byte) error {
+	// Declare a private type to avoid recursion into this function.
+	type createCredentialRequest CreateCredentialRequest
+
+	var rawMessage json.RawMessage
+	dec := createCredentialRequest{
+		Proof: &rawMessage,
+	}
+
+	err := json.Unmarshal(b, &dec)
+	if err != nil {
+		return err
+	}
+	if dec.Type == nil {
+		return ErrMissingField("type")
 	}
 
-	if len(req.Verifier) == 0 {
-		return ErrInvalidVerifier
+	switch *dec.Type {
+	case CredentialTypeAWSSTS:
+		dec.Proof = &CredentialProofAWSSTS{}
+	case CredentialTypeRSA:
+		dec.Proof = &CredentialProofRSA{}
+	default:
+		return ErrInvalidCredentialType
 	}
+	if rawMessage != nil {
+		err = json.Unmarshal(rawMessage, dec.Proof)
+		if err != nil {
+			return err
+		}
+	}
+	*req = CreateCredentialRequest(dec)
+	return nil
+}
 
+// Validate validates the request fields.
+func (req *CreateCredentialRequest) Validate() error {
+	if req.Fingerprint == nil {
+		return ErrMissingField("fingerprint")
+	}
+	if req.Verifier == nil {
+		return ErrMissingField("verifier")
+	}
+	if req.Type == nil {
+		return ErrMissingField("type")
+	}
 	err := req.Type.Validate()
 	if err != nil {
 		return err
 	}
+	if *req.Type == CredentialTypeAWSSTS && req.Proof == nil {
+		return ErrMissingField("proof")
+	}
+	fingerprint, err := CredentialFingerprint(*req.Type, req.Verifier)
+	if err != nil {
+		return err
+	}
+	if *req.Fingerprint != fingerprint {
+		return ErrInvalidFingerprint
+	}
+
+	return nil
+}
+
+// CredentialProofAWSSTS is proof for when the credential type is AWSSTS.
+type CredentialProofAWSSTS struct {
+	Region  *string `json:"region"`
+	Request []byte  `json:"request"`
+}
 
+// Validate whether the CredentialProofAWSSTS is valid.
+func (p CredentialProofAWSSTS) Validate() error {
+	if p.Region == nil {
+		return ErrMissingField("region")
+	}
+	if p.Request == nil {
+		return ErrMissingField("request")
+	}
 	return nil
 }
+
+// CredentialProofRSA is proof for when the credential type is RSA.
+type CredentialProofRSA struct{}
+
+// CredentialFingerprint returns the fingerprint of a credential.
+func CredentialFingerprint(t CredentialType, verifier []byte) (string, error) {
+	var toHash []byte
+	if t == CredentialTypeRSA {
+		// Provide compatibility with traditional RSA credentials.
+		toHash = verifier
+	} else {
+		encodedVerifier := base64.RawStdEncoding.EncodeToString(verifier)
+		toHash = []byte(fmt.Sprintf("credential_type=%s;verifier=%s", t, encodedVerifier))
```

Comment:
Technically, yes, you're right, it should work. But since we're using a string format as a "container", it is nice to keep everything a string. 

Not applicable in this case, but applying the strategy you suggest can in some cases lead to collision attacks. For example, if we had the following structure: `a=<value-of-a>;b=<value-of-b>`, I can think of a few value for `a` and `b` that produce a collision. For example:

a: `1;b=2`
b: `3`

Produces the same result as:
a: `1`
b: `2;b=3`

Both would produce `sha256(`a=1;b=2;b=3`)` as a result. Not good.

In our case `credential_type` is limited to a few values. But still it is a good practice to not open up for these kinds of things. That also why (though not strictly necessary), I used `base64.RawStdEncoding`. This omits the `=` as padding.